File: algorithms/cpp/maxNumberofK-SumPairs/maxNumberofK-SumPairs.cpp
// Source: https://leetcode.com/problems/max-number-of-k-sum-pairs/
// Author: Miao Zhang
// Date:   2021-05-27

class Solution {
public:
    int maxOperations(vector<int>& nums, int k) {
        unordered_map<int, int> m;
        int res = 0;
        for (int num: nums) m[num]++;
        for (int num: nums) {
            if (m[num] < 1 || m[k - num] < 1 + (num + num == k)) continue;
            m[num]--;
            m[k - num]--;
            res++;
        }
        return res;
    }
};

File: algorithms/python/maxNumberofK-SumPairs/maxNumberofK-SumPairs.py
#! /usr/bin/env python3
# -*- coding: utf-8 -*-

# Source: https://leetcode.com/problems/max-number-of-k-sum-pairs/
# Author: Miao Zhang
# Date:   2021-05-27

class Solution:
    def maxOperations(self, nums: List[int], k: int) -> int:
        cnt = collections.Counter(nums)
        res = 0
        for num in nums:
            if cnt[num] < 1 or cnt[k - num] < 1 + (num + num == k):
                continue
            cnt[num] -= 1
            cnt[k - num] -= 1
            res += 1
        return res
